include/revng/Yield/TagType.h

```diff
@@ -17,6 +17,7 @@ type: enum
   - name: MnemonicSuffix
   - name: Register
   - name: Whitespace
+  - name: Target
```

Comment:
I agree with it being vague. It was intended, I wasn't trying to put too much meaning into it preliminarily as I'm still unsure how it's going to turn out, I didn't want to cut possibilities by choosing a restrictive name.

Also I knew you are not going to like how vague it is and that this discussion was coming.

1. This is close to what I opted out for on the llvm side (it also includes calls, but I don't feel like it's a big deal).
2. We could go with this option, too. It would require a bit more of llvm-side redesign (since I didn't mark non-jump instructions - it's not like we can embed too much information in about the non-code parts of the binary). But this is also a valid option we might want.

> I'd like this to be reflected in LLVM too

I would be **very** opposed to contradicting terminology. All of llvm tags have an obvious counterpart in this enum (I did extend the abbreviations though, because I don't like them too much) and I'd **really** prefer it stayed that way.

> What do we want to use this for?

I feel like this is the question in question (pun intended). Can we squeeze anything useful out of non-code addresses? If so, we should definitely go with the second option. Otherwise the first one sounds good enough for me.

Reply:
OK, let's talk about this.

In theory, we could avoid `<target>` entirely and just go with `<imm>` inspecting if the content is a valid address for the current model.
However, imagine a binary loaded at address `0`, where small numbers might be valid addresses. In that case, we'd like the disassembler to hint at us that a particular immediate is *for sure* an address.

```asm
example1: ; No way to distinguish
  mov rax, 0x1000

example2: ; For sure an address!
  lea rax, [rip+0x1000]

example3: ; For sure an address!
  jmp 0x1000
```

Can you report on how the LLVM disassembler currently reports each of these (both in absolute and relative modes)?

Once we have these situations in front of us we can better decide which route to take.

Reply:
I'm definitely opposed to inspecting every immediate. That's just asking for trouble. The options are see either "we only try to be smart about jump/call targets" or "we try to be smart about all memory accesses, like loads, stores, you get the point".

> Can you report on how the LLVM disassembler currently reports each of these

Everything is just an immediate. But as you saw, for most architecture, it's pretty trivial to patch the codegen to redirect some instruction "groups" to be printed by any code we want. So we have enough freedom to do anything we want, I'd say.

Reply:
Can you report on how the LLVM disassembler currently reports each of the above (both in absolute and relative modes)?

In particular, I'd like to see how `rip+0x1000` is handled. Is it collapsed into an absolute value or does `rip` always remain.

Reply:
Ah, right. Sorry, I didn't notice you singling `rip` out. Yeah, that's the fascinating side of the problem I didn't think of.

```c
// x86-64 assembly
  lea rdi, [rip + 0xf79]
//***----------------------mnemonic
//----***------------------register
//---------*************---memory
//----------***------------register
//----------------****-----immediate
```

Reply:
> does `rip` always remain.

I have never seen it getting collapsed. I guess it's up to us :sadparrot:

Reply:
As for the rest, `absolute` mode
```
// x86-64 assembly
  jmp 0x2120
//***----------mnemonic
//----******---immediate
//----******---target
```

in immediate mode:
```c
// x86-64 assembly
  jmp 0xfffffffffffffd10
//***----------------------mnemonic
//----******************---immediate
//----******************---target
```

Reply:
There's an extended discussion on mattermost that eclipses this thread, so I'll close it.